tokio-tasks/src/task_map.rs

```diff
@@ -0,0 +1,70 @@
+use crate::future_ext::FutureExt;
+use futures::future::RemoteHandle;
+use futures::Future;
+use std::collections::HashMap;
+use std::hash;
+
+/// Struct controlling the lifetime of the async tasks, such as
+/// running actors and periodic notifications. If it gets dropped, all
+/// tasks are cancelled.
+///
+/// Each task is associated with a key, so that it can be removed from
+/// the internal `HashMap` whenever consumers want to stop tracking
+/// the corresponding task. By removing it from the `HashMap` we free
+/// up space.
+pub struct TaskMap<K>(HashMap<K, RemoteHandle<()>>);
+
+impl<K> TaskMap<K>
+where
+    K: Eq + hash::Hash,
+{
+    /// Spawn the task on the runtime and remember the handle.
+    ///
+    /// The task will be stopped if this instance of [`TaskMap`] goes
+    /// out of scope.
+    pub fn add(&mut self, key: K, f: impl Future<Output = ()> + Send + 'static) {
+        let handle = f.spawn_with_handle();
+        self.0.insert(key, handle);
+    }
+
+    /// Spawn a fallible task on the runtime and remember the handle.
+    ///
+    /// The task will be stopped if this instance of [`TaskMap`] goes
+    /// out of scope. If the task fails, the `err_handler` will be
+    /// invoked.
+    pub fn add_fallible<E, EF>(
```

Comment:
```suggestion
    pub fn insert_fallible<E, EF>(
```

🙃 Might be more idemoatic

Reply:
insert is an implementation detail, IMHO we shouldn't leak out how we internally deal with futures that finish 😇

Reply:
But the fact that we replace what was inside is *not* an implementation detail! That is what is indicated by insert. Add would mean that it is added.

Reply:
Sorry I missed this discussion.

I chose to use the same names as for `Tasks` because I obviously copy-pasted 99% of this code :rofl:

I think it' s an important comment though, because it highlights something that I hadn't considered. The point of introducing this type is to allow us to `remove` tasks that have finished, not to overwrite tasks that might still be ongoing! With the underlying `HashMap` it seemed natural to allow overwriting (current state of the PR), but we could decide that `add` and `add_fallible` should do nothing if the corresponding key is present in the map. What do you think, @klochowicz @da-kami?

The naming discussion is interesting too. I do think that we don't need to reveal the fact that we use a `HashMap` internally by using the exact same terminology, but it may indicate that we are overwriting in case of a collision, making the API easier to understand. Maybe we can find a name that is both descriptive and not too influenced by the underlying type.

Reply:
I think we should consider *not* allowing overwrites. I think we should ensure that a task finishes (we can add auto-timeouts at some point if we want, but that's a different feature) and not just kill a task.

So yeah, I think we should guard that we only add if the ID is not present - and then `add` also makes sense.

Reply:
As discussed with @klochowicz, this PR doesn't actually allow ephemeral tasks to be overwritten.

The `TaskMap` abstraction does allow the user to overwrite tasks, but because the `spawner::Actor` generates a _unique_ ID for every task, tasks are never actually overwritten.

Let's move forward with this PR and we can come back to the naming discussion later (or not if @klochowicz's alternative solution is adopted eventually).